### ISSUE TITLE
LS441DE

### ISSUE THREAD

**Issue by jcmoore1977** [1000001101000/Debian_on_Buffalo#8]
Still having network issues. Sorry It has been a while. I had to replace my hard drive PCB. It took a crash.
I have used the updated files you put on here, now all the network does is blink then shuts off. I can not even start installer. The zip file I received from you worked for a moment, then went back to shutting off. Any suggestions. I have no stock firmware on drive it is blank and formatted with the files in /

**Comment by 1000001101000** [comment]
I've made a number of improvements to the ls441 dtb since we last worked on it. I learned that the ls400 devices use one of the pins from the ethernet chip (normally meant for led3) to control whether the device shuts down or restarts. This might have explained the problems you were seeing (maybe). 

Now that i've accounted for that and a few other things i'm hoping we'll have better luck.

I'm a little unclear about how far you're getting right now, it sounds like it's failing to boot at all. 

which installer are you currently trying?
what are the leds on the front doing when you try?
how big is the partition you're trying to boot from?

**Comment by jcmoore1977** [comment]
Let me see if I can make a video and share it with you. Give me about 20 mins or so.

**Comment by jcmoore1977** [comment]
Here is a photo of the files I use. The one highlighted and the initrd at the top.

https://1drv.ms/u/s!AsfCHdLfE-658Q9Z1SxgfL_fQmc-

**Comment by jcmoore1977** [comment]
Here is the video of what the system does upon power on. I went back and forth from front to back where the network is plugged in. At the end you can see all lights solid and net work is off.

https://1drv.ms/v/s!AsfCHdLfE-658RBbtqLihW6wK2QZ

**Comment by jcmoore1977** [comment]
Partition in about 200MB on boot

**Comment by jcmoore1977** [comment]
I get this error when I can finally get into installer shell and type exit

Jan  1 00:00:30 sshd[1041]: Server listening on :: port 22.
Jan  1 00:00:30 apt-install: Queueing package openssh-server for later installation
Jan  1 00:00:37 kernel: [   37.785266] mvneta d0074000.ethernet eth0: Link is Down
Jan  1 00:00:39 kernel: [   39.862199] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
Jan  1 00:00:59 kernel: [   59.625323] mvneta d0074000.ethernet eth0: Link is Down
Jan  1 00:01:00 kernel: [   60.662152] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
Jan  1 00:01:28 sshd[1056]: Accepted password for installer from 192.168.0.13 port 65123 ssh2
Jan  1 00:01:28 sshd[1057]: lastlog_openseek: Couldn't stat /var/log/lastlog: No such file or directory
Jan  1 00:01:28 sshd[1057]: lastlog_openseek: Couldn't stat /var/log/lastlog: No such file or directory
Jan  1 00:01:28 debconf: Setting debconf/language to en
Jan  1 00:02:21 kernel: [  141.785336] mvneta d0074000.ethernet eth0: Link is Down
Jan  1 00:02:23 kernel: [  143.862244] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
Jan  1 00:02:26 kernel: [  146.985336] mvneta d0074000.ethernet eth0: Link is Down
Jan  1 00:02:29 kernel: [  149.062189] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
Jan  1 00:02:49 kernel: [  169.865259] mvneta d0074000.ethernet eth0: Link is Down
Jan  1 00:02:51 kernel: [  171.942247] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
Jan  1 00:03:00 kernel: [  180.265201] mvneta d0074000.ethernet eth0: Link is Down
Jan  1 00:03:01 kernel: [  181.302203] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
Jan  1 00:03:14 kernel: [  194.825204] mvneta d0074000.ethernet eth0: Link is Down
Jan  1 00:03:15 kernel: [  195.862269] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
Jan  1 00:03:18 kernel: [  198.985257] mvneta d0074000.ethernet eth0: Link is Down
Jan  1 00:03:20 kernel: [  200.022213] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
Jan  1 00:03:18 kernel: [  198.985257] mvneta d0074000.ethernet eth0: Link is Down

**Comment by 1000001101000** [comment]
alright, so we're using the current stretch installer (that's what i'd recommend).
it looks like it's booting, the light changing to solid indicates it booted something.

the ethernet led turns off at the point i'd expect it to turn solid, I've got it configured as follows:
LED0: On - Link, Blink - Activity, Off - No Link

I'm going to try with mine and confirm what mine does (it's been a while since I looked at it). I want to be sure the current files are working for me and the led behavior is how I remember it before we get too far.

did you ever end up getting a usb nic?

**Comment by 1000001101000** [comment]
oh, nice. that's helpful. looks like the link keeps going up and down for some reason.

back when we first worked on this that was probably triggering the device to shutdown too.... at least we're past that issue.

**Comment by jcmoore1977** [comment]
Yes I have order one. It will be here before the end of the month. When HDD board went out my concern was getting everything off drive first.

**Comment by 1000001101000** [comment]
understandible.

have you tried using a different ethernet cable/switch/etc?

**Comment by jcmoore1977** [comment]
Yes, I have recently replace all the network devices cables and switches in my entire house. The stock firmware still works perfectly.

**Comment by 1000001101000** [comment]
I figured you had but I thought I'd check. I checked the logs on a couple of my devices and didn't see any disconnections like yours. The fact that the stock firmware seems stable makes me hopeful there is some kind of solution though the fact that my ~10 devices don't do this nor do the devices of anyone else I've interacted with make me think there is something physically different about your device.

I'm going to do some research and see if there is a setting to make the driver and/or ethernet chip less sensitive to disconnects or something of that nature. I'll also take a look at the buffalo kernel source code again (I remember they had the code for the whole led2 as restart indicator thing, I don't recall what else they customized).

There are a few things you could try/look at:
1. You could try the buster installer. It's possible the newer mvneta driver handles this better
2. you could take a look at the logs when running the stock firware and confirm whether you see those disconnects or not
3. you could try pinging the device continously for a while and see if packets are getting dropped and what kind of latency you get.

one other thing to note: I recently changed the installer configuration to auto detect the network adapter rather than just assume it will be eth0 (I've added support for the ts3000 series which have 2 ethernet ports). This should help when you try the usb3 adapter but also means it might not detected the on-board ethernet if it thinks the link is down when the auto detect runs.

**Comment by jcmoore1977** [comment]
yeah buster doesn't work at all. I have been reading up on this even though it is for my current kernel 3.3.4 stock.

I see some things are the same and some are different. Your thoughts.

/ SPDX-License-Identifier: GPL-2.0
/*
 * Device Tree common file for the Seagate Personal Cloud NAS 1 and 2-Bay
 * (Armada 370 SoC).
 *
 * Copyright (C) 2015 Seagate
 *
 * Author: Simon Guinot <simon.guinot@sequanux.org>
 */

/*
 * TODO: add support for the white SATA LED.
 */

#include "armada-370.dtsi"
#include <dt-bindings/gpio/gpio.h>
#include <dt-bindings/input/input.h>

/ {
	chosen {
		stdout-path = "serial0:115200n8";
	};

	memory@0 {
		device_type = "memory";
		reg = <0x00000000 0x20000000>; /* 512 MB */
	};

	soc {
		ranges = <MBUS_ID(0xf0, 0x01) 0 0xf1000000 0x100000
			  MBUS_ID(0x01, 0xe0) 0 0xfff00000 0x100000>;

		internal-regs {
			coherency-fabric@20200 {
				broken-idle;
			};

			serial@12000 {
				status = "okay";
			};

			ethernet@74000 {
				status = "okay";
				pinctrl-0 = <&ge1_rgmii_pins>;
				pinctrl-names = "default";
				phy = <&phy0>;
				phy-mode = "rgmii-id";
			};

			usb@50000 {
				status = "okay";
			};
		};
	};

	regulators {
		compatible = "simple-bus";
		#address-cells = <1>;
		#size-cells = <0>;

		regulator@0 {
			compatible = "regulator-fixed";
			reg = <0>;
			regulator-name = "USB Power";
			regulator-min-microvolt = <5000000>;
			regulator-max-microvolt = <5000000>;
			regulator-always-on;
			regulator-boot-on;
			gpio = <&gpio1 27 GPIO_ACTIVE_LOW>;
		};
		regulator@1 {
			compatible = "regulator-fixed";
			reg = <1>;
			regulator-name = "SATA0 power";
			regulator-min-microvolt = <5000000>;
			regulator-max-microvolt = <5000000>;
			enable-active-high;
			regulator-always-on;
			regulator-boot-on;
			gpio = <&gpio1 18 GPIO_ACTIVE_HIGH>;
		};
	};

	gpio-keys {
		compatible = "gpio-keys";
		#address-cells = <1>;
		#size-cells = <0>;

		power {
			label = "Power button";
			linux,code = <KEY_POWER>;
			gpios = <&gpio1 19 GPIO_ACTIVE_HIGH>;
			debounce-interval = <100>;
		};
		reset {
			label = "Reset Button";
			linux,code = <KEY_RESTART>;
			gpios = <&gpio1 23 GPIO_ACTIVE_LOW>;
			debounce-interval = <100>;
		};
		button {
			label = "USB VBUS error";
			linux,code = <KEY_UNKNOWN>;
			gpios = <&gpio1 21 GPIO_ACTIVE_LOW>;
			debounce-interval = <100>;
		};
	};

	gpio-leds {
		compatible = "gpio-leds";

		red-sata0 {
			label = "cumulus:red:sata0";
			gpios = <&gpio1 26 GPIO_ACTIVE_HIGH>;
			default-state = "off";
		};
	};

	gpio_poweroff {
		compatible = "gpio-poweroff";
		gpios = <&gpio1 25 GPIO_ACTIVE_HIGH>;
	};
};

&pciec {
	status = "okay";

	/* USB 3.0 Bridge ASM1042A */
	pcie@1,0 {
		status = "okay";
	};
};

&mdio {
	pinctrl-0 = <&mdio_pins>;
	pinctrl-names = "default";

	phy0: ethernet-phy@0 {
		reg = <0>;
	};
};

&pinctrl {
	pinctrl-0 = <&sata_led_pin>;
	pinctrl-names = "default";

	sata_led_pin: sata-led-pin {
		marvell,pins = "mpp60";
		marvell,function = "sata0";
	};
	gpio_led_pin: gpio-led-pin {
		marvell,pins = "mpp60";
		marvell,function = "gpio";
	};
};

&spi0 {
	status = "okay";
	pinctrl-0 = <&spi0_pins2>;
	pinctrl-names = "default";

	spi-flash@0 {
		#address-cells = <1>;
		#size-cells = <1>;
		/* MX25L8006E */
		compatible = "mxicy,mx25l8005", "jedec,spi-nor";
		reg = <0>; /* Chip select 0 */
		spi-max-frequency = <50000000>;

		partition@0 {
			label = "u-boot";
			reg = <0x0 0x100000>;
		};
	};

**Comment by 1000001101000** [comment]
1. What do you mean when you say buster doesn’t work at all? I’ve run buster on mine for a while without any particular issues.

2. That’s a device tree for another armada370 device you can look at them all in the kernel source under arch/arm/boot/dts. You can find information about the various device bindings in /Documentation

You could certainly try adjusting the ethernet settings by modifying the device tree though I dont know of anything specific to try (i’d offer it up if I did).

The datasheet for the ethernet chip is available too, if you find a setting you want to change you can set it using phytool and make it permanent via adding it to the device tree.

Unfortunately I can’t help much with that line of reasearch, i’ve got 15 different devices (i just counted) running that ethernet chip and none are having this problem.

**Comment by 1000001101000** [comment]
I’ve got a link to the datasheet for the ethernet chip and all the reset of the hardware info I put together here:
https://buffalonas.miraheze.org/wiki/Linkstation_LS441D

If you want an example of manually setting registers on the chip check out the ls400 restart script in the tools section.

You can see how to set that same register in the device tree under the ethernet binding

**Comment by jcmoore1977** [comment]
Buster will boot device, but the network shuts off immediately. Can’t even login to it. 
It’s just very strange how stock version runs perfectly. 2 days now and no network shut down. I wish I could use the original buffalo uimage and then boot the stretch initrd file

**Comment by 1000001101000** [comment]
yeah. I want to know why your device behaves like this when none of the others seem to and I also want to know what's different about the stock firmware that doesn't trigger it.

I'd really like to see your dmesg output from the stock firmware to see if it's happening but the software handles it better or if their kernel just somehow avoids the issue.

you could probably get phytool running on the stock firmware and use it to dump all the register values which we could then compare to the debian values for a comparison.

you could also try running Jessie and see if the 3.16 kernel works any better for you.

you could also try running jessie with the buffalo kernel, I did with my ls421de for a while when I was first experimenting with this stuff. you can see an example of how to create an initrd for that type of setup in this project:
https://github.com/rogers0/OpenLinkstation

**Comment by 1000001101000** [comment]
one other thing i'd be curios about is whether your device might have a different revision of main board or something like that. If you feel like taking your apart you can compare to some fairly high resolution pictures of mine at:
https://buffalonas.miraheze.org/wiki/Linkstation_LS441D

**Comment by jcmoore1977** [comment]
Okay I will do the things you suggest later this afternoon. I wil definitely take apart so we can see the board

**Comment by 1000001101000** [comment]
nice!
I think verifying whether the board is different will be good, though I don't really know what kind of differences to look for, definitely make sure to get a good picture of the lettering on the Ethernet chip. 

The biggest question I have is whether you'll see those constant up/down messages in the logs of the stock firmware or not. If so then buffalo's firmware must have a way of handling them more gracefully and we can try to something similar. 

If not then it's probably something more subtle about how the device is initialized or how that older driver worked. If so you should be able to dump a bunch of info and register values using ethtool and phytool (though getting those working under the stick firmware could be challenging).

**Comment by jcmoore1977** [comment]
Plus my usb network stick will be here by this weekend. So we can dig in for sure then.

**Comment by jcmoore1977** [comment]
Here are this photos of my MB. 
https://1drv.ms/f/s!AsfCHdLfE-658RooP1DS_BNKluou

**Comment by 1000001101000** [comment]
Hmm one is a -ca and one is a -cb. I have no idea if that means anything but it’s interesting. Definitely the same ethernet chip.

**Comment by jcmoore1977** [comment]
Yes all the components looked the same. 
My adapter should arrive tomorrow, so I will be digging into this, this weekend. I will post some dmesg on stock and if I can get Debian loser to stay active post those as well

**Comment by jcmoore1977** [comment]
Okay before I plug in the debian installer here is the complete dmesg from stock firmware.
[dmesg-stock.txt](https://github.com/1000001101000/Debian_on_Buffalo/files/2999476/dmesg-stock.txt)

I also ran the dmesg command 10 mins later and it is the same thing as the above post.

**Comment by jcmoore1977** [comment]
So far no issues with network, using usb nic. I have been up and running for 10 mins now.

**Comment by jcmoore1977** [comment]
Also is there a reason why your its files are different for stretch vs buster?

**Comment by 1000001101000** [comment]
nice!
I suppose the fact that that is stable lets us further narrow down the issue to the built in nic rather than something that coincidentally caused the link to go up and down.

your dmesg output seems to confirm that the problem isn't occuring in the stock firmware. That raises more questions than it answers but it's good to know. something is different about your device than all the others we've seen and whatever that difference is the stock firmware is fine with... somehow.

There are a few directions I can think of looking but they get progressively more complicated and I'm not sure what our best bet is.

I think using ethtool to look at all the network settings both under debian and stock might show if they're setting some sort of option differently. 

if that doesn't give us any hints I could write a script using phytool to dump all the register values in the ethernet chip and we could compare them.

**Comment by 1000001101000** [comment]
There have been a number of changes to the device-tree specification and to the different bindings which require the device-trees to be updated to match the kernel they go with.

This makes keeping up with the testing branch a pain, I'm not sure if i'll keep posting testing images like I have been with Buster or not. Hopefully there will be less changes to account for this time around.

**Comment by jcmoore1977** [comment]
Okay, I have installed the stretch software using the usb nic. I want to let it run for a bit to see what happens with this installed. Then I will use to built-in network. I want to test a theory I have.

**Comment by jcmoore1977** [comment]
dimes with usb nic plugged in.
[message.txt](https://github.com/1000001101000/Debian_on_Buffalo/files/2999606/message.txt)

**Comment by 1000001101000** [comment]
theories are good. mine are getting harder and harder to test.

I should have told you to get a usb3 gigabit adapter. The one you got will be good for testing (and good for testing other non-usb3 devices) but won't be a good replace ent for the built in one..... if it came to that.

**Comment by jcmoore1977** [comment]
Yes, I do have another NAS by a completely different company. I use it to hold all my stuff on. This is just something to mess around with in my spare time.

But that theory did not work. I tried to use the system with built in nic. But all lights go off all I can to is use the usb one.

**Comment by jcmoore1977** [comment]
this is something that sticks out in system log..

Mar 23 11:39:03 debian systemd-udevd[195]: Could not generate persistent MAC address for eth0: No such file or directory

**Comment by 1000001101000** [comment]
I see that in my logs too, I think it's just because it's not getting the mac address from the chip or the boot loader.

**Comment by jcmoore1977** [comment]
Okay, the built in network comes on when system is booting up kernel, but upon finishing boot it turns off.

**Comment by 1000001101000** [comment]
What does your interfaces file look like?

**Comment by jcmoore1977** [comment]
networkctl
WARNING: systemd-networkd is not running, output will be incomplete.

IDX LINK             TYPE               OPERATIONAL SETUP     
  1 lo               loopback           n/a         unmanaged 
  2 eth0             ether              n/a         unmanaged 
  3 enx00e04c534458  ether              n/a         unmanaged 


and from dmesg...

[    2.564836] mvneta d0074000.ethernet eth0: Using random mac address 92:71:46:7b:b8:44
[   76.236581] usb 2-1: Product: USB 2.0 10/100M Ethernet Adaptor
[   76.397538] dm9601 2-1:1.0 eth1: register 'dm9601' at usb-d0050000.usb-1, Davicom DM96xx USB 10/100 Ethernet, 00:e0:4c:53:44:58

**Comment by 1000001101000** [comment]
No, that’s “normal” for these devices.

The device gets a random mac early in the boot process because it can’t get it from the adapter and hasn’t loaded /etc/network/interfaces yet.

Your usb adapter has the mac address stored in the ethernet chip which the system can look up and use right away. 

Let me zip up some logs from a device and post them. I’m thinking anything suspicious you find in your logs wouldn’t be in mine which should help

**Comment by jcmoore1977** [comment]
Okay trying something else. I have the led for network staying on. I am adjusting the clock settings in the its file to see if these will help.

**Comment by jcmoore1977** [comment]
okay that did not work 2.573125] OF: /soc/internal-regs/ethernet@74000: could not find phandle
[    2.601463] mvneta: probe of d0074000.ethernet failed with error -2

but the light is on..

**Comment by 1000001101000** [comment]
What exactly did you change?

**Comment by jcmoore1977** [comment]
ethernet@70000 {
				reg = <0x70000 0x4000>;
				interrupts = <0x8>;
				clocks = <0x7 0x4>;
				status = "disabled";
				compatible = "marvell,armada-370-neta";
			};

			mdio {
				#address-cells = <0x1>;
				#size-cells = <0x0>;
				compatible = "marvell,orion-mdio";
				reg = <0x72004 0x4>;
				clocks = <0x7 0x4>;

				ethernet-phy@0 {
					reg = <0x0>;
					marvell,reg-init = <0x3 0x10 0xf000 0x881>;
					linux,phandle = <0x8>;
					phandle = <0x8>;
				};
			};

			ethernet@74000 {
				reg = <0x74000 0x4000>;
				interrupts = <0xa>;
				clocks = <0x7 0x3>;
				status = "okay";
				compatible = "marvell,armada-370-neta";
				phy = <0x8>;
				phy-mode = "rgmii-id";

to this

ethernet@70000 {
				reg = <0x70000 0x4000>;
				interrupts = <0x8>;
				clocks = <&gateclk 4>;
				status = "disabled";
				compatible = "marvell,armada-370-neta";
			};

			mdio {
				#address-cells = <0x1>;
				#size-cells = <0x0>;
				compatible = "marvell,orion-mdio";
				reg = <0x72004 0x4>;
				clocks = <&gateclk 4>;

				ethernet-phy@0 {
					reg = <0x0>;
					marvell,reg-init = <0x3 0x10 0xf000 0x881>;
					linux,phandle = <0x8>;
					phandle = <0x8>;
				};
			};

			ethernet@74000 {
				reg = <0x74000 0x4000>;
				interrupts = <0xa>;
				clocks = <&gateclk 3>;
				status = "okay";
				compatible = "marvell,armada-370-neta";
				phy = <0x8>;
				phy-mode = "rgmii-id";
			};


Just look at the clocks info. basically put <&gateclk 4>; or <&gateclk 3>;

**Comment by 1000001101000** [comment]
On the bright side your dtb worked

**Comment by jcmoore1977** [comment]
Yeah, I changed it to 4 on the ethernet@74000, now it’s in a boot loop. I need to plug hard up to my Linux box and put the backup of uImage, and initrd so I can boot up and go back to drawing board. So it has something to do with the line that starts with ethernet@74000. 

I may go back into your archives and get the old dts to compared

**Comment by 1000001101000** [comment]
There’s another dtb you can compare to here:
https://sites.google.com/site/shihsung/88fxxxx-soc

Plus you can look at any of the other mainline armada370/xp ones

**Comment by jcmoore1977** [comment]
Yeah I was on that site last night doing some research. I am going to try some things after work today.

**Comment by jcmoore1977** [comment]
Okay I found the buffalo kernel here. 

http://opensource.buffalo.jp/ls400-110.html

Now to look and see how they bind some of their devices.

**Comment by 1000001101000** [comment]
I went ahead and wrote the script I was talking about to dump the control register values for the ethernet chip. I ran it against my ls421de and uploaded the output. you can find those under:
Tools/physcan.sh
Tools/ls421de_stretch_phy.txt

You can see what the values mean from the datasheet found here:
https://www.marvell.com/documents/eoxwrbluvwybgxvagkkf/

I'm thinking we should try running the script on the stock firmware (which might require tweaking) and compare the values. If we find a relevant difference we can test ii by changing the corresponding value using phytool and if we find something that solves your problem we could change it permanently in the device tree.

That would only work if the difference in behavior between the stock firmware and Debian is related to how the chip is initialized/configured rather than some kind of logic within the driver/kernel.

**Comment by jcmoore1977** [comment]
I ran you above script.
I get this error

error: phy_write (-524)
error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)
 error: phy_read (-524)

and when running lshw I get this error

*-network:0 DISABLED
       description: Ethernet interface
       physical id: 1
       logical name: eth0
       serial: fe:ac:35:ac:36:db
       capabilities: ethernet physical
       configuration: broadcast=yes driver=mvneta driverversion=1.0 link=no multicast=yes
  *-network:1
       description: Ethernet interface
       physical id: 2
       logical name: enx00e04c534458
       serial: 00:e0:4c:53:44:58
       size: 100Mbit/s
       capacity: 100Mbit/s
       capabilities: ethernet physical tp mii 10bt 10bt-fd 100bt 100bt-fd autonegotiation
       configuration: autonegotiation=on broadcast=yes driver=dm9601 driverversion=22-Aug-2005 duplex=full firmware=Davicom DM96xx USB 10/100 Ether ip=192.168.0.38 link=yes multicast=yes port=MII speed=100Mbit/s

of course ethernet interface 1 is my usb nic.

**Comment by 1000001101000** [comment]
did you run it under debian or under the stock firmware?
I believe the interface needs to be up for it to work.

**Comment by jcmoore1977** [comment]
Debian,
But the network will not stay up. 
It cuts off that why I am using the usb nic

So I don’t know if it’s a setting I need to change or what.

**Comment by 1000001101000** [comment]
Hmm, it probably depends on your interface settings whether it tries to bring it back up after it goes down.

We can use the data I got from my device for the debian example. I can probably get the stock firmware example too if I get a chance to load it in the next couple days.

**Comment by jcmoore1977** [comment]
Okay, I will look into why it may go down and not back up

**Comment by jcmoore1977** [comment]
okay after running this command the eth0 comes up, but not talk to network.
I am however able to run your pyscan.sh now and get this.
[pyscan.txt](https://github.com/1000001101000/Debian_on_Buffalo/files/3011014/pyscan.txt)

So I am wondering if it gets turned off at boot time or not.
any ideas on what I should look for. I know the usual place ( /etc/network/interface) file. It will be tomorrow be I can check again.

**Comment by jcmoore1977** [comment]
Wondering if I could build the network module from the Buffalo kernel on the Debian kernel and insert it. Or maybe look at the source code differences?

**Comment by 1000001101000** [comment]
You can download the regular 3.3.4 source and do a diff to see the differences in the buffalo source and see if you find something helpful. You can also compare those drivers to the modern equivalents.

I've looked into it a bit but am not sure I've understood it properly. 

It looks like this ethernet chip is currently supported by the sky2 module and mvneta module. mvneta appears to specifically for marvel armada SoCs where sky2 is used elsewhere (The intel-based Terastations use the sky2 driver for the same chip for example). It looks like back in 3.3.4 days it was all just the sky2 driver (i think). I tried unloading the mvneta module and loading the sky2 module but it didn't seem to detect the device. It could be that the modern sky2 driver doesn't scan for devices using that type of connection or that you'd have to add info to the device tree to make that work (like compatible=sky2 or something).

If you do find a use useful difference you can try compiling that into a new module and give it a try. Oddly enough I'm doing something similar to patch a driver on another project, specifically build-modules.sh from:
https://github.com/1000001101000/Debian_on_Intel_Terastations

I had to patch the IT87 drivers to add GPIO support for the specific chip being used by these devices so that I could control the front panel buttons. 

I'm still hoping if you find something it's something we can just add to the DTB but if it does require patching a module I can add that as an option for anyone who runs into the same problem (so far you seem to be the only one).

**Comment by jcmoore1977** [comment]
Here are the files that have to do with net that are different..

Files linux-3.3.4-buffalo/drivers/net/Makefile and linux-3.3.4/drivers/net/Makefile differ
Files linux-3.3.4-buffalo/drivers/net/ppp/ppp_generic.c and linux-3.3.4/drivers/net/ppp/ppp_generic.c differ
Files linux-3.3.4-buffalo/drivers/net/ppp/pppoe.c and linux-3.3.4/drivers/net/ppp/pppoe.c differ

I could try and build these under debian and see what would happen.

**Comment by 1000001101000** [comment]
as far as I know ppp/pppoe are a protocol used for dsl connections. I have a hard time imagining the relavence, though they raises the wuestion of what they actually changed.

what do the actual differences look like?

I think some of the differences will be from a generic patch provided by marvel which may include stuff for routers which could explain it if it doesn't look relavent.

**Comment by 1000001101000** [comment]
as far as I know ppp/pppoe are a protocol used for dsl connections. I have a hard time imagining the relavence, though they raises the wuestion of what they actually changed.

what do the actual differences look like?

I think some of the differences will be from a generic patch provided by marvel which may include stuff for routers which could explain it if it doesn't look relavent.

**Comment by jcmoore1977** [comment]
I can post differences later today, I am also wondering, isn’t the driver for that call net.ko or Ethernet.ko, or something along those lines?

**Comment by 1000001101000** [comment]
probably, the various kernel drivers/modules depend on each other at multiple levels. I know from working on this that the armada370/xp use mvneta and mvmdio to talk to the ethernet controller. I'm less certain of anything upstream from there. 

I can see from the stock dmesg output that the sky2 driver is loaded at boot in the stock firmware but I'm not entirely certain if it depends on something else or how that all works. I suspect some of what's currently handled by mvneta/mvmdio is in the mach-mvebu stuff somewhere.

**Comment by jcmoore1977** [comment]
I found this out

The configuration item CONFIG_SKY2:

prompt: Marvell Yukon 2 support
type: tristate
depends on: CONFIG_PCI
defined in drivers/net/Kconfig
found in Linux kernels: 3.1
modules built: sky2, sky2

**Comment by 1000001101000** [comment]
I may be confusing things talking about the sky2 driver, it may be specific to PCI. I only bring it up because I can see the driver starts in the stock firmware:
`sky2: driver version 1.30`
`i2c /dev entries driver`
`rtc-rs5c372 0-0032: rs5c372a found, 24hr, driver version 0.6`

On my Terastation runnig debian I see that the driver is actually used for the ethernet adapters:
`[    1.451298] sky2: driver version 1.30`
`[    1.451696] sky2 0000:03:00.0: Yukon-2 Optima chip revision 1`
`[    1.453053] sky2 0000:03:00.0 eth0: addr 4c:e6:76:28:46:05`
`[    1.453631] sky2 0000:02:00.0: Yukon-2 Optima chip revision 1`
`[    1.454769] sky2 0000:02:00.0 eth1: addr 4c:e6:76:28:46:04`
`[    1.552658] sky2 0000:03:00.0 enp3s0: renamed from eth0`
`[    1.576475] sky2 0000:02:00.0 enp2s0: renamed from eth1`
`[   10.805897] sky2 0000:02:00.0 enp2s0: enabling interface`
`[   10.937362] sky2 0000:03:00.0 enp3s0: enabling interface`
`[   13.350565] sky2 0000:02:00.0 enp2s0: Link is up at 1000 Mbps, full duplex, flow control both`

I wonder if the driver provided by marvel is only for 2.6.xx is because that is the last version before they got it included in the mainline linux kernel. 

I should probably note that the ethernet chip is a marvell alaska chip rather than yukon, though it seems the drivers are the same for both.

**Comment by jcmoore1977** [comment]
Do you happen to know how to get the original DTB from stock firmware. I looked in the arch/arm/boot/dtbs folders in the Buffalo kernel I downloaded, but I could not find it. I am wondering just for cross reference something.

**Comment by 1000001101000** [comment]
The stock kernel doesn't use DTBs, the method they use to define the hardware pre-dates that method.

Instead most of that info is defined under the mach-mvebu and plat-mvebu directories and the various other custom things they added to the kernel.

**Comment by jcmoore1977** [comment]
That will give me a start point.

**Comment by jcmoore1977** [comment]
I have been looking at the Buffalo kernel. I see info on the mv643xx but nothing on the 74000 you are using. Do you think there is a difference in hardware there

**Comment by 1000001101000** [comment]
I believe mv643xx refers to a different chip that was used on older devices.

70000 & 74000 refer to the memory addresses used to access the on-board ethernet devices. You can see more info about them looking at armada370.dtsi and armada370xp.dtsi. That might help with your comparison to the stock kernel since that’s where a lot of that info is on the modern kernel

**Comment by 1000001101000** [comment]
I finally got around to loading phytool on the stock firmware, unfortunately I couldn't get it to work there.

**Comment by jcmoore1977** [comment]
I think I fixed it. Network is up and running running a ping test to see if I lose packets. I will report back. Let me see how log this will stay up before I get to excited. But at least the port is open and I am talking to box without usb nic

**Comment by jcmoore1977** [comment]
Never mind. It shuts off after a few mins then stays off or will come back on but can not talk to box. Have to pull power cord. I am going to plug drive up to linux box and look at sys logs. see if there is something in those.

**Comment by 1000001101000** [comment]
I spent a little time looking at your phytool output compared to mine. Most of the Page 0 differences appear to be related to the fact that your link was down when it read that register where mine was up. There are some interesting failure counters and things of that nature that might be interesting to look at over time to see if your chip is registering the disconnects or if it's purely in the OS but I don't think that gets any closer to a solution.

Page 1 is all fiber stuff which I assume doesn't apply. I'm surprised there are as many differences but this could also be related to the link being detected as down I suppose.

Page 3 shows the expected LED setting difference between the ls421 and ls441 device trees. It also shows a polarity invert bit set for an led which could mean you were testing that other dtb or could be normal for the ls421 vs ls441 from some uboot setting. 

Page 5 shows a different crossover reading on your device than mine. I assume this is from out device/switches doing something different with autocrossing or just an artifact of your link being down at the time. 

Page 6 shows you have different values in your chips temperature sensor than mine. No surprise there.

most of the other differences were undocumented or uninteresting. overall I didn't see anything particularly helpful. 

It's too bad I wasn't able to get the values from the stock firmware. I still kind of hope that we could discover that the buffalo firmware disables some autonegotiation setting in the chip (or something like that) that once identified we could test on your device by manually setting it with phytool and then if it fixed your problem set it permanently in the dtb.

maybe I should spend some time trying to re-build phytool against the 3.3.4 kernel source and see if that fixes it.

**Comment by styelz** [comment]
I have the same problem, the nic responds to ping after having been booted up for about 5 minutes.
Then it turns off after about 2 minutes.
I can ssh to the IP when it's up, but I don't know what the root password is ? Can someone tell me please ? I've tried the default, "password"

**Comment by jcmoore1977** [comment]
If your trying to install you should ssh as installer and password install, if you have already installed. The password would be what ever you set it as durning install. 
I have used a usb nic card. It has no issues. But I would still like to figure out why it turns off.

**Comment by styelz** [comment]
Thanks. Damn, yea it disconnects me part way through the install,I'd like to get this running I was hoping to use BTRFS with it. I've tried changing the firmware from ls400-v185 to ls400-v173 with no luck though.Not sure if that makes any difference either. I will try get it working on a USB nic, hopefully i can tether via my phone or something. Will let you know if I find out anything useful.

**Comment by 1000001101000** [comment]
I imagine you get a USB3 Gigabit adapter and get similar performance as the on-board ethernet. 

I wouldn't think downgrading the the firmware would accomplish anything but it was certainly worth a try. I've made sure mine was flashed with the latest firmware/boot loaded when testing to ensure that works. 

I've personally installed Debian on 2 different ls441de's without this issue and have heard from a few others who have also had success on theirs. Including all the other installs myself and others have done on other ls200/ls400 devices (all using the same ethernet chip with the same soc) it adds up to dozens of devices using the same hardware/drivers without issue.

It's interesting that we've found a second device with the same issue now, i wonder if this one is the same hardware revision as @jcmoore1977 . 

I haven't run across anything new that would explain this issue but there are a couple of things you could look at:
1. you could try a newer kernel and see if any of the recent changes improve anything. I just installed Debian's 5.2 kernel on my TS1400D (closest thing to the ls441de I have online right now) and it worked just fine.
https://packages.debian.org/sid/linux-image-5.2.0-2-armmp

2. I noticed while trying to get WOL working on this device that Raspbian has some custom patches they maintain for their kernel, some of which affect this network driver. You could try recompiling with one of these patches or even try installing a Raspbian kernel and see if you have better luck. 
https://github.com/armbian/build/blob/master/patch/kernel/mvebu-next/

3. I suppose you could also try filing a bug with the maintainers of the mvneta/mvmdio driver and see they have any ideas for moving forward.

**Comment by styelz** [comment]
I checked the motherboard and the MVLS441D-CA version.

I saw you mentioned something about the LED of the ethernet may be the cause, I found some info here which I guessing you have already seen.

GPIO pin 15 is connected to both the power button and pin 8 (LED[2]) of the ethernet chip. Initially I assumed this had something to do with Wake-On-LAN but this device doesn't claim to support WOL. Presumably this is used to signal to the bootloader whether the device intends to shutdown or reboot when it shuts down. This can cause troubling behavior under linux if you do not change the default behavior of the LED[2] pin and leave the power button configured.

I am waiting on my USB NIC to arrive so I can take a closer look.

**Comment by 1000001101000** [comment]
Looks like a quote from the wiki page I wrote on the subject. I was able to infer that from a kernel patch someone else put together to handle the LED2 behavior on his device, the Buffalo firmware  handles it in pretty much the same way. You can find that patch here:
https://sites.google.com/site/shihsung/88fxxxx-soc

I've added similar logic to the DTB to make sure LED2 is disabled until needed. Instead of patching the kernel I created a script which enables LED2 at the end of the restart process to allow the device to reboot.

When this issue was originally reported every time the link dropped it was triggering LED2 which triggered the power button and shutdown the device. I've changed the button to send KEY_ESC rather than KEY_POWER in addition to disabling LED2 which  solves that part of the problem.

I've still got no idea why the link shows as dropping for both of your devices. Does your device work properly in the stock firmware the same as @jcmoore1977 ?

**Comment by styelz** [comment]
Hi, thanks for the links. The device works fine with stock firmware.

**Comment by 1000001101000** [comment]
Sounds like we know that whatever the issue is it can be corrected by software. It also seems like it might be correlated by different hardware revisions.

I believe we've confirmed both of your boards are MVLS441D-CA and mine is MVLS441D-CB.

I wonder if the difference is something we'd be able to see. Here are some reasonably high-res pictures of my board. The ethernet chip is visible on the back:
https://static.miraheze.org/buffalonaswiki/d/da/441_board1.jpg
https://static.miraheze.org/buffalonaswiki/4/48/441_board2.jpg

If the chips are different we might be able to find documentation specific to them that helps us come up with a solution. 

can you take a look at yours and maybe grap a closeup of the area around the ethernet chip for comparison?

**Comment by styelz** [comment]
Sure, here's the images.
https://i.imgur.com/RbasJTO.jpg
https://i.imgur.com/7mLhV4X.jpg

**Comment by 1000001101000** [comment]
Are you able to attach them? I’m having trouble getting full size from imgur

**Comment by 1000001101000** [comment]
Nevermind, desktop site worked better.

I don’t see anything obvious about the ethernet chip or it’s visible connections.

**Comment by styelz** [comment]
I plugged the USB-NIC in and was able to connect ok. I saw in syslog the interface was going up and down over and over. The interface shows as down, so I did an "ip link set dev eth0 up" and a "ip addr add 192.168.0.25 dev eth0" and was able to connect in on that IP. The link stayed up and ping didnt fail at all but i could still see the syslog message saying the nic was going down and back up. When i did an "ip addr" it showed eth0 with "NO-CARRIER" and then "BROADCAST" when the syslog messages occurred.

~ # ip a

2: eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc mq qlen 1024
    link/ether dc:fb:02:da:02:fb brd ff:ff:ff:ff:ff:ff

~ # ip addr add 192.168.0.25/24 dev eth0
~ # ip link set dev eth0 up
~ # ip a

2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq qlen 1024
    link/ether dc:fb:02:da:02:fb brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.25/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::defb:2ff:feda:2fb/64 scope link
       valid_lft forever preferred_lft forever

~ # ip a

2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq qlen 1024
    link/ether dc:fb:02:da:02:fb brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.25/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::defb:2ff:feda:2fb/64 scope link
       valid_lft forever preferred_lft forever

Syslog Entries when nic shows NO-CARRIER

Aug 17 11:16:06 kernel: [ 2639.821032] mvneta d0074000.ethernet eth0: PHY [d0072004.mdio-mii:00] driver [Marvell 88E1510]
Aug 17 11:16:06 kernel: [ 2639.821364] mvneta d0074000.ethernet eth0: configuring for phy/rgmii-id link mode
Aug 17 11:16:06 kernel: [ 2639.821579] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
Aug 17 11:18:20 kernel: [ 2774.021107] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
Aug 17 11:18:20 kernel: [ 2774.021139] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
Aug 17 11:18:35 kernel: [ 2788.360227] mvneta d0074000.ethernet eth0: Link is Down
Aug 17 11:18:38 kernel: [ 2791.429103] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
Aug 17 11:18:43 kernel: [ 2796.552221] mvneta d0074000.ethernet eth0: Link is Down
Aug 17 11:19:06 kernel: [ 2820.101209] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
Aug 17 11:19:10 kernel: [ 2824.200275] mvneta d0074000.ethernet eth0: Link is Down

**Comment by styelz** [comment]
Not sure if it's important, but when I type "reboot" it shutsdown and I have to manually power it back on.

**Comment by 1000001101000** [comment]
hmmm. I wonder how performance is when it’s doing that. 

Can you try pinging it from another device and see if packets drop?

If it’s performing normally, just randomly reporting carrier drops we might be able to disable the carrier check or something. I’ll take another look at the chip’s datasheet when I get a chance.

**Comment by styelz** [comment]
No drops with pings at all. It's also strange when I checked eth0 with ethtool it shows it in half duplex mode. 

# ethtool eth0
Settings for eth0:
        Supported ports: [ TP MII FIBRE ]
        Supported link modes:   10baseT/Half 10baseT/Full
                                100baseT/Half 100baseT/Full
                                1000baseT/Full
        Supported pause frame use: Symmetric
        Supports auto-negotiation: Yes
        Supported FEC modes: Not reported
        Advertised link modes:  1000baseT/Full
        Advertised pause frame use: Symmetric
        Advertised auto-negotiation: Yes
        Advertised FEC modes: Not reported
        Speed: 1000Mb/s
        Duplex: Half
        Port: MII
        PHYAD: 0
        Transceiver: internal
        Auto-negotiation: on
        Supports Wake-on: g
        Wake-on: d
        Link detected: no

I did this to fix that:
# ethtool -s eth0 speed 1000 duplex full autoneg off

**Comment by styelz** [comment]
Now I have the main OS installed it looks like it tries to bring the interface up 3 times and then stops trying and leaves it as down. Initially when I bring it up, it shows as ok, then drops to NO-CARRIER.
I tried setting the device as the default gateway device but it would not work for some reason. It's like the route is blocked due to the nic reporting as down.
 
[    7.409161] mvneta d0074000.ethernet eth0: Using random mac address 56:9e:d5:2d:80:ca
[   47.192811] mvneta d0074000.ethernet eth0: PHY [d0072004.mdio-mii:00] driver [Marvell 88E1510]
[   47.201993] mvneta d0074000.ethernet eth0: configuring for phy/rgmii-id link mode
[   47.214739] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[   52.420929] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control rx/tx
[   52.429031] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   65.735020] mvneta d0074000.ethernet eth0: Link is Down
[   69.827739] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control off
[   87.239256] mvneta d0074000.ethernet eth0: Link is Down
[   90.308054] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control off
[   94.407277] mvneta d0074000.ethernet eth0: Link is Down
[   97.476072] mvneta d0074000.ethernet eth0: Link is Up - 1Gbps/Full - flow control off
[  101.575237] mvneta d0074000.ethernet eth0: Link is Down

**Comment by 1000001101000** [comment]
Hmm. Does that mean those stable pings were actually to the usb nic?

**Comment by jcmoore1977** [comment]
When I used my usb nic, it worked flawless. I even had the on board working for 10 mins on time. Then is stopped. I was going to add this line to dtb and retry the boot. 

marvell,reg-init = <3 16 0xff00 0x1e 3 17 0xfff0 0x00>;

This is from this web site.   https://forum.trenz-electronic.de/index.php?topic=791.0

From reading. Wondering if the phy register gets rewritten upon loading the driver

**Comment by 1000001101000** [comment]
Those look like the led registers, i know for a fact the DTB settings override the driver defaults. 

If you want to try changing their values you can using phytool, you can then make them permanent by changing the dtb. For an example of how to do this check out my ls400_restart script.

I have a link to the datasheet which describes the registers on the wiki:
https://buffalonas.miraheze.org/wiki/Linkstation_LS441D

Note that this device has registers divided into pages, when you switch to page 3 to edit the led registers the network will drop, to restore it you must switch back to page 0. This means you need to issue them all as a one-liner or script otherwise the network won’t come back.

**Comment by styelz** [comment]
I did some more testing and it looks like it must be going thruough the USB-NIC somehow to get to the eth0 IP. If i boot with just the onboard nic it's back to up and down. not stable like i though. Runs great with the USB nic.

**Comment by styelz** [comment]
Not sure how to use phytool. I tried following your example in the script without any luck.
Most of this is new to me btw. Looks like you set the page first and then the register ?

Also noticed the output of 'ip link' is missing "LOWER_UP" which means there is no physical cable attached ?

BTW, thank you very much for your help and the images for the NAS, it's working great with the USBNIC.

**Comment by 1000001101000** [comment]
This should display the registers @jcmoore1977 was talking about:

phytool write eth0/0/22 3; phytool read eth0/0/16; phytool read eth0/0/17; phytool write eth0/0/22 0

The interface does need to be up for it to work.

**Comment by styelz** [comment]
First time I ran it, it gave this:
0x1881
0x4400

Subsequent calls give this:
0x101e
0x4400

**Comment by 1000001101000** [comment]
That’s interesting. The first result is the value I set in the dtb, the second looks like the driver default.

**Comment by styelz** [comment]
I tried kernel 5.2.0.2, still the same. Is there a way I can dump the nics registry values to a file ? I tried ethtool -d but the card doesn't allow it. I could try compare the debian one to stock of so.

Also, When looking at the [datasheet](https://github.com/1000001101000/Debian_on_Buffalo/issues/8), it mentions "bits".  Do I convert the binary number i want to set to hex and write that with phytool ?

**Comment by jcmoore1977** [comment]
Did you try to change this line in the dtb found here to this

marvell,reg-init = <0x3 0x10 0xf000 0x0881> —original
 
marvell,reg-init = <3 16 0xff00 0x1e 3 17 0xfff0 0x00> —to this. 

I am going to start here. I believe I will have some free time this weekend.

**Comment by 1000001101000** [comment]
the format of those DTB entries is:
<page register mask value ....>
**note that the "original" entries are all hexadecimal but the other mixes decimal and hexadecimal, ie 0x10 == 16 etc.

You can see what they represent in the datasheet on page 108 - 110:
https://www.marvell.com/documents/eoxwrbluvwybgxvagkkf/

you should see that my DTB sets LED 1 and 2 to 0x8 (OFF) and sets LED 0 as 0x1 (On - Link, Blink - Activity, Off - No Link)

My restart script changes LED2 to 0x9 (ON) right before restart to signal a restart vs a shutdown.

**Comment by 1000001101000** [comment]
I have noticed newer dtbs have started including this:
`marvell,fiber = "disabled";`
It might be worth trying.

**Comment by 1000001101000** [comment]
@styelz 
> Is there a way I can dump the nics registry values to a file 
Yes, I created a script for this. It's in the tools directory "physcan.sh".

**Comment by styelz** [comment]
I can't get phytool to run on stock :( I tried many different combinations on eth1 but all give this result:

[root@LS441D2FB ~]# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 16436 qdisc noqueue
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
2: eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop qlen 512
    link/ether 00:50:43:f5:1d:87 brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP20> mtu 1500 qdisc pfifo_fast qlen 512
    link/ether dc:fb:02:da:02:fb brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.25/24 brd 192.168.0.255 scope global eth1

[root@LS441D2FB ~]# phytool write eth1/0/22 3
error: phy_write (-95)

**Comment by styelz** [comment]
Also, when I do a shutdown on the debian install, it restarts. It works fine if I do a halt -p

**Comment by 1000001101000** [comment]
I wasn’t able to get it to work on stock either. It might be possible to recompile it using the stock kernel source/headers and maKe it work but it might just be that phytool didn’t work on the older kernels or that the older driver is incompatible in some way.

Whether it shutdown/restart will be driven by that led2 pin, if your values are changing as shown in your earlier dump it is likely set as a “link” light which would explain that behavior. 

Before I got the shutdown behavior sorted I used halt too, still do one some devices.

**Comment by styelz** [comment]
I found I could run ethtool -d on stock, not sure how to translate it though.

[root@LS441D2FB /]# ethtool -d eth1 raw off
Offset          Values
------          ------
0x0000:         00 01 01 00 00 00 00 00 00 10 00 02 00 00 00 00
0x0010:         38 00 00 01 00 00 00 00 01 00 00 00 00 00 00 00
0x0020:         05 00 00 00 01 00 11 00 00 00 00 00 00 00 00 00
0x0030:         00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0x0040:         00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0x0050:         00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0x0060:         00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0x0070:         00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

**Comment by 1000001101000** [comment]
I would guess that is one of the "pages" of register values from the chip. if so I would expect you would get the same output from phytool for whichever page it is. 

It might also be a combination that the driver puts together and makes available to ethtool.

**Comment by jcmoore1977** [comment]
As you can see I have both links up and running for about 5 mins now.
eth0 and my usb nic enx00e04c534458
I will let it sit here while I watch some college football.
Will report back soon

 eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group defa                                                               ult qlen 1024
    link/ether dc:fb:02:da:03:93 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::defb:2ff:feda:393/64 scope link
       valid_lft forever preferred_lft forever
3: enx00e04c534458: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast                                                                state UP group default qlen 1000
    link/ether 00:e0:4c:53:44:58 brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.20/24 brd 192.168.0.255 scope global dynamic enx00e04c534458
       valid_lft 85828sec preferred_lft 85828sec
    inet6 fe80::2e0:4cff:fe53:4458/64 scope link
       valid_lft forever preferred_lft forever

**Comment by jcmoore1977** [comment]
I added this line to the /etc/network/interfaces file

# The primary network interface
allow-hotplug eth0
iface eth0 inet dhcp
hwaddress DC:FB:02:DA:03:93
allow-hotplug enx00e04c534458
iface enx00e04c534458 inet dhcp

**Comment by jcmoore1977** [comment]
that did not work, I have been doing some reading on this matter. There are patched in phy.c file I am in the process of rebuild the phy module. I will post my findings.

Like you said in the above comments it could be a software issues seeing how our boards are different Model A vs Model B

This is the readme from the patch I am trying.

Configuring the PHY LED registers for the Marvell 88E1510 and others is
not possible, because regardless of the values in marvell,reg-init, it
is later overridden in m88e1121_config_aneg with a non-standard default.

This patch moves that default configuration to .config_init to allow
setting the LED configuration through marvell,reg-init in the device
tree, which should override said default if it exists.

**Comment by jcmoore1977** [comment]
Okay it is something on the driver/kernel side of things. The card is fine on the uboot side (ex. DTB). so with some time and patching of the marvel.c file (this is what resets the card from device tree settings) I may be able to get this up for good.

**Comment by 1000001101000** [comment]
I took a brief look at the kernel source code published by buffalo for this device and looked specifically at what they changed from the vanilla 3.3.4 source:
https://opensource.buffalo.jp/ls400-110.html
https://mirrors.edge.kernel.org/pub/linux/kernel/v3.x/

It looks like a large part of it is just a patch provided by Marvell adding support for the SoC and associated components which we're not part of the mainline kernel yet. 

They appear to use mvEthPhyRegRead and mvEthPhyRegWrite to write the PHY registers in the kernel code. You can see that calls they use when they initialize/manage the device in:
arch/arm/plat-armada370/mv_drivers_lsp/mv_network/mv_ethernet/mv_ethernet.c
arch/arm/plat-armada370/mv_drivers_lsp/mv_network/mv_ethernet/mv_eth_tool.c
arch/arm/plat-armada370/mv_drivers_lsp/mv_network/mv_ethernet/mv_gateway.c
arch/arm/plat-armada370/mv_drivers_lsp/mv_proc/proc.c
arch/arm/plat-armada370/mv_drivers_lsp/mv_switch/mv_switch.c
arch/arm/plat-armada370/mv_hal/eth-phy/mvEthPhy.c
arch/arm/plat-armada370/mv_hal/eth-phy/mvEthPhy.h

It might be worth comparing how some of those registers were set compared to how they are set by the modern driver.

You can even see where they use it to set that LED2 register at shutdown in arch/arm/kernel/process.c
void machine_halt(void)
 {
+#if defined(CONFIG_BUFFALO_PLATFORM) && defined(CONFIG_ARCH_ARMADA370)
+	mvEthPhyRegWrite(0, 0x00000016, (u16)0x0003);
+	mvEthPhyRegWrite(0, 0x00000010, (u16)0x1880);
+	msleep(10);

**Comment by jcmoore1977** [comment]
Thanks, on a side note I did change one value in your device tree. The part where you set register as 0x00 on the phy side to 0x01. This kept the device link on. Now I just need to hook it up along with the usb nic to see if link is up or down. And why it does not communicate. I don’t know if it needs to have a register of 1 or the default you have as 0. I will try to do some more digging. I believe I am in the right track, it’s just a matter of time. If I am successful I can let you know the changes just in case anyone else has the same board or problems we are facing.

**Comment by 1000001101000** [comment]
cool.

If you come up with a solution I'll gladly add it to the installer whether it's a DTB change or some sort of startup script or default configuration change.

**Comment by jcmoore1977** [comment]
Correct me if I am wrong here. On you DTB is see your led code is written to 0xF000 which equals 15, but shouldn’t that code be written to reg 17 which is 0x1100. From reading up on the data sheet. Also another data sheet for the 88E1111 shows it to be on reg 24 which should be 0x1800

**Comment by 1000001101000** [comment]
It's a 88E1518 which has a different register layout.

Here is the correct datasheet:
https://www.marvell.com/documents/eoxwrbluvwybgxvagkkf/

I've also got a link to it on: 
https://buffalonas.miraheze.org/wiki/Linkstation_LS441D

**Comment by jcmoore1977** [comment]
Sure, FYI it is from 4 years ago though, looks like it needs to be patched within the uboot, so I will have to look into it that way. 

https://u-boot.denx.narkive.com/O5Lxe0fa/patch-0-2-net-support-marvell-88e1510-and-fix-88e151x-init

**Comment by 1000001101000** [comment]
I don't think the bootloader is the problem, if it were you'd see the same problems in the stock firmware.

Which setting are you looking to try, is it this part?
+ /* Select page 18 */
+ phy_write(phydev, MDIO_DEVAD_NONE, 22, 18);
+
+ /* In reg 20, write MODE[2:0] = 0x1 (SGMII to Copper) */
m88e1518_phy_writebits(phydev, 20, 0, 3, 1);

- /* Phy reset */
+ /* PHY reset is necessary after changing MODE[2:0] */
m88e1518_phy_writebits(phydev, 20, 15, 1, 1);
- phy_write(phydev, MDIO_DEVAD_NONE, 22, 0); /* reg page 18 */
+
+ /* Reset page selection */
+ phy_write(phydev, MDIO_DEVAD_NONE, 22, 0);
+

**Comment by jcmoore1977** [comment]
I have tried all I know, Have been at this for over 24 hours. Time to call it for now. I just don't understand why it would boot up then link up, link down, link up, link down, then stay down. I am exhausted now. All I can say is looking at the data sheets the 88e1510 and 88e1518 are almost identical except for the voltage. I really need to understand why the stock firmware handles this prefect and the debian on our boards fails. Something is happening after boot up with the kernel. I am just not sure where to look.

**Comment by 1000001101000** [comment]
yeah, trying to dig through kernel source is a pain. especially when you’re not sure what you’re after. What i’d really like to try is adding some code to the stock kernel to dump the phy registers so we can compare them ( either that or get phytool to work on stock). easier said than done though.

I just remembered something, I saw in the buffalo source that it disables auto-negotiation at one point (I think). I though it might be worth trying.

`ethtool -s eth0 autoneg off speed 1000 duplex full`

also did you ever try adding this to the dtb:
`marvell,fiber = "disabled";`

I don’t know why but newer device trees have been adding that. I you like I can build one for you try out.

**Comment by jcmoore1977** [comment]
I did boot up the drive with the stock kernel. Found out it is using the mv_eth driver for the net connection. It also have eth0 and eth1 as loaded. I don’t know enough about device trees to add another pay, but I will look into the mv_eth file and I am wondering if there is a bridge somewhere. I see you disabled the ether@700000 but have ether@74000 enabled. Next time I boot my Linux laptop I can post the ethtool output from the stock kernel

**Comment by 1000001101000** [comment]
That’s an interesting thought. I wouldn’t think enabling the other port would accomplish anything (there’s no second chip or port there) but it is on the stock kernel. It’s something to try

**Comment by jcmoore1977** [comment]
I also see it is using RMII where you tree sets it as RGMII. I know that is for gigabyte. Also I have noticed the usb nic I uses deletes the /dev/class/eth0 or eth1. So even if I did get it working properly (a miracle) I would more than likely have to reinstall everything. When kernel boots it tells you ex??????? (USB nic) has renamed that.

**Comment by jcmoore1977** [comment]
May I ask on your chip m88e1518 is it rev A or B. 
My chip reads 88E1518-NWB2

**Comment by 1000001101000** [comment]
I don’t think you need to worry about what the usb nic does, if you remove it from the interfaces file it won’t do any of that. 

I think you can see the markings of the chip on this picture:
https://static.miraheze.org/buffalonaswiki/4/48/441_board2.jpg

If you look around the wiki you can see the markings on the other devices using that same chip too. 

As far as grmii vs rmii etc, I don’t really know much about. Most of those entries are based on the other armada370 dts in the linux kernel.

**Comment by jcmoore1977** [comment]
We have the exact same chip. I am just wondering something about the difference in CA and CB boards now. 

Also do you know much about busybox devmem

**Comment by 1000001101000** [comment]
I've used busybox quite a bit over the years but have never used devmem. I've never really done anything with direct memory read/write. All the real direct writing stuff I've done has been over MDIO/I2C/SPI/ISA etc and I've used a tool for that in each case (phytool, i2cset etc).

**Comment by jcmoore1977** [comment]
I was able to dump my mud locks 0-4 and just glanced at them with hex editor. There is some interesting stuff in there. Once I look further into this I will post some output of these. Maybe there is a configuration in these (address and offset)

**Comment by 1000001101000** [comment]
mud locks?

**Comment by jcmoore1977** [comment]
Lol MTDBlocks 0-4

**Comment by 1000001101000** [comment]
ah okay. 
mtdblock0 contains the boot loader image (u-boot) 
mdtblock1 contains the u-boot environment variables (run fw_printenv to see them) 
mtdblock2+ are the NAND flash containing the images the device uses to boot when there are not disks and well as a copy of the stock firmware which it uses to set up disks when first used.

**Comment by 1000001101000** [comment]
I generally don't mess with the mtdblock devices because you can render the device unbootible if you make a mistake. In that case recovery requires a serial console which is pretty hard to set up on these devices (requires de-soldering a chip or something like that).

If you do decide to make changes you'll need to remove the read-only setting I added to the device tree. 

That said "shihsung" uses a custom uboot image for his device:
https://sites.google.com/site/shihsung/88fxxxx-soc/ls441de

He's included the patch he uses for ubott which includes some code for setting up the ethernet devices:
+void reset_misc (void)
+{
+	struct udevice *eth;
+	u16 phyaddr;
+	u16 reg;
+
+	eth = eth_get_dev();
+	if (!eth) {
+		printf("%s: can't get ETH device\n", __FUNCTION__);
+		return;
+	}
+
+	if (miiphy_read(eth->name, 0xEE, 0xEE, &phyaddr)) {
+		printf("%s: can't read PHY device address\n", __FUNCTION__);
+		return;
+	}
+
+	miiphy_write(eth->name, phyaddr, 0x16, 0x0003);
+	miiphy_read(eth->name, phyaddr, 0x10, &reg);
+	reg &= ~0x0f00;
+	reg |= 0x0900;
+	miiphy_write(eth->name, phyaddr, 0x10, reg);
+	miiphy_write(eth->name, phyaddr, 0x16, 0x0000);
+
+	mdelay(10);
+}

**Comment by jcmoore1977** [comment]
O no, I wouldn’t dare think to write any data to these areas. I did the DD command and saved them to a flash drive for studies only.

**Comment by 1000001101000** [comment]
If you're just interested in the environment variables give fw_printenv a try, it should work properly for that model. Some of the other models have all their variables embedded which typically leaves me using string/hexdump to look at them too.

**Comment by jcmoore1977** [comment]
I believe the issue is with the mvneta.ko module. I have unloaded it and the lights stay on for the on-board nic card. Now I will have to recompile that kernel module using the source code. BTW I am on stretch. Last time I tried to load buster I could not. I believe they updated Buster so the modules will not load.

**Comment by jcmoore1977** [comment]
I am done. Don’t know what else to do. Look like this NAS will go back to storage for now. I’m exhausted on this matter. Never have I had so much problems with a network device.

**Comment by 1000001101000** [comment]
sorry for the delayed reply.

Did you do a dist-upgrade to get to buster or use my installer?

I wish I was able to re-create the issue so I could be more helpful. 
Did you ever get a chance to try either of these:

> I just remembered something, I saw in the buffalo source that it disables auto-negotiation at one >point (I think). I though it might be worth trying.
>
>ethtool -s eth0 autoneg off speed 1000 duplex full
>
>also did you ever try adding this to the dtb:
>marvell,fiber = "disabled";

**Comment by jcmoore1977** [comment]
I have tried the Marvell, fiber disabled, but with the Ethernet not working the ethtool will not work on eth0. It will tell you no such device exists or it will say read error.

**Comment by 1000001101000** [comment]
You can probably get it to execute the moment it comes up using something similar to
`ifup eth0 && ethtool -s eth0 .....`
If that were to work we could make it an if-up script or make it part of the dtb (somehow)

**Comment by jcmoore1977** [comment]
I used ip link set dev eth0 up, then try ethtool command, that’s when I get the read error. It does seem like it is an issue with the way the kernel handles the nic card. I have even put the hwaddress in the interfaces file with the correct MAC address. Like I say I am at a loss of ideas.

**Comment by 1000001101000** [comment]
Did you ever end up trying Debian Jessie on the device? If so you can ignore the rest of this:

I ran across something today, have you seen it?
https://www.spinics.net/lists/arm-kernel/msg559146.html

It sounds like the same issue and reports that it only started in 4.2. would you be willing to install a 4.1 kernel on stretch and see if it becomes stable? If so I can take a look at the changes between 4.1 and 4.2.

The 4.1 kernel package can be found here:
https://snapshot.debian.org/archive/debian/20150912T095543Z/pool/main/l/linux-latest/linux-image-armmp_4.1%2B67~bpo8%2B1_armhf.deb

**Comment by 1000001101000** [comment]
This email form the beginning of that thread is the one that seems familiar:
https://www.spinics.net/lists/arm-kernel/msg558959.html

**Comment by jcmoore1977** [comment]
I did run across something like that. I suppose I could adjust your build script to get Jessie instead of stretch or buster build the image and see if that would work. Unless you wanted to build a 4.1 uImage and I could use that with the INITRD installer

**Comment by jcmoore1977** [comment]
I see you already have Jessie files here, are those built upon kernel 4.1

**Comment by 1000001101000** [comment]
They are based on 3.16 I believe. Those should at least boot, possibly still work to do a Jessie install (it depends on what Debian is still hosting).

The other option would be to install with the regular stretch installer and then install that 4.1 package with dpkg

**Comment by jcmoore1977** [comment]
I will those images a shot, remember the Ethernet did not even stay up for the installer, so if it works just to play around with the installer that is progress, and we know it is something to do with kernel then.

**Comment by jcmoore1977** [comment]
I have found the issue, It is with the kernel. I have my onboard nic card up and running for now.

I am so excited.

mvneta d0074000.ethernet eth0: link up

I did not patching either. I have some ideas to see if it will stay up in the installer as well.

**Comment by 1000001101000** [comment]
Nice! 
Is that by running Jessie or something else?

**Comment by jcmoore1977** [comment]
Settings for eth0:
        Supported ports: [ TP MII ]
        Supported link modes:   10baseT/Half 10baseT/Full
                                100baseT/Half 100baseT/Full
                                1000baseT/Half 1000baseT/Full
        Supported pause frame use: No
        Supports auto-negotiation: Yes
        Advertised link modes:  100baseT/Full
        Advertised pause frame use: No
        Advertised auto-negotiation: No
        Speed: 100Mb/s
        Duplex: Full
        Port: MII
        PHYAD: 0
        Transceiver: external
        Auto-negotiation: off
        Link detected: yes

**Comment by jcmoore1977** [comment]
I am running jessie, but I bet this will work on all others as well
Just with I knew how to extract the install initrd file to make the one change.

**Comment by jcmoore1977** [comment]
It has to do with the autoneg when the kernel boots up.

**Comment by jcmoore1977** [comment]
I added the following line to the rc.local file after I installed it with my usb nic card.

ethtool -s eth0 autoneg off speed 100 duplex full
exit 0

Of course you have to install ethtool first as it is not installed by the installer.

**Comment by jcmoore1977** [comment]
But It's still up and running for 10 mins now, and not one issue. Working just like stock firmware now

**Comment by jcmoore1977** [comment]
from mii-tool
eth0: 100 Mbit, full duplex, link ok
  product info: vendor 00:50:43, model 29 rev 1
  basic mode:   100 Mbit, full duplex
  basic status: link ok
  capabilities: 1000baseT-HD 1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
  advertising:  100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD

**Comment by jcmoore1977** [comment]
going to try and update to latest jessie kernel first. Then on to stretch.

**Comment by 1000001101000** [comment]
sounds like a plan.

also try 1000M.

**Comment by jcmoore1977** [comment]
It's doesn't like that ethtool -s eth0 autoneg off speed 1000 duplex full.

It shuts down the whole network on that.

**Comment by jcmoore1977** [comment]
When I do that, it goes right back to the whole issue we have been having and why this is an open issue.
I will leave it at 100 for now.

**Comment by 1000001101000** [comment]
interesting. 

Stable on 100M seems like a good start, hopefully we can find a way to get it to full speed eventually (or confirm stock does the same I guess)

**Comment by jcmoore1977** [comment]
yeah no idea on that.

**Comment by jcmoore1977** [comment]
The new kernel is working. Let continue on with stretch

**Comment by jcmoore1977** [comment]
Still up and running on buster.

Settings for eth0:
        Supported ports: [ TP MII FIBRE ]
        Supported link modes:   10baseT/Half 10baseT/Full
                                100baseT/Half 100baseT/Full
                                1000baseT/Full
        Supported pause frame use: Symmetric
        Supports auto-negotiation: Yes
        Supported FEC modes: Not reported
        **Advertised link modes:  100baseT/Full**
        Advertised pause frame use: Symmetric
        **Advertised auto-negotiation: No**
        Advertised FEC modes: Not reported
        Speed: 100Mb/s
        Duplex: Full
        Port: MII
        PHYAD: 0
        Transceiver: internal
        Auto-negotiation: off
        Supports Wake-on: g
        Wake-on: d
        Link detected: yes

So still have it set to 100 and autoneg off

**Comment by 1000001101000** [comment]
awesome, that's a big step forward. 

Can you post the steps to set it up so that @styelz and @RQWorldblender can give it a try?

**Comment by jcmoore1977** [comment]
You have to install ethtool.
1.) apt-get install ethtool
2.) check in /etc for a file called rc.local If not present do nano rc.local and copy the following in to it

#!/bin/sh -e
# This script is executed at the end of each multiuser runlevel

ethtool -s eth0 autoneg off speed 100 duplex full
exit 0

Then save and exit.

3.) chmod 755 rc.local
4.) reboot and link should stay up.

This is a work around for now, but at least the on board link stays up.

**Comment by 1000001101000** [comment]
looks like ethtool would be pretty easy to embed in the installer:
~# ldd /usr/sbin/ethtool
	linux-vdso.so.1 (0xbeec7000)
	libm.so.6 => /lib/arm-linux-gnueabihf/libm.so.6 (0xb6e6e000)
	libc.so.6 => /lib/arm-linux-gnueabihf/libc.so.6 (0xb6d74000)
	/lib/ld-linux-armhf.so.3 (0xb6f3b000)

Likewise setting up that script would also be fairly simple. If you can find a combination that allows it to work as 1000M I could see adding that to the process.

**Comment by 1000001101000** [comment]
I suppose we could try to make a script that runs at startup and automatically downgrades to 100M if it detects the problem. I'd like some solution that doesn't require another network adapter, I just don't want to mess up devices not experiencing the problem.

**Comment by 1000001101000** [comment]
could you try the mtu/100M/autoneg separately? Id be curious to know what the smallest possible change ends up being.

**Comment by jcmoore1977** [comment]
well every time I change it to speed 1000 it shuts the network off and I have to reboot.

**Comment by jcmoore1977** [comment]
I changed to mtu only to stop getting RX packet drops, but this is for my case. Others my leave default of 1500.

**Comment by 1000001101000** [comment]
right on.  I wonder what the output of ethtool on the stock firmware looks like.

**Comment by 1000001101000** [comment]
In any case I'm going to start working on adding ethtool to the installer (and adding it as a package that gets installed). I'm going to need to do some work/testing on them for some other work I'm doing so it's a good time for that.

**Comment by jcmoore1977** [comment]
Do you know if you use phytoplankton to write to register is that saved after reboot or does it go back to defaults after a shutdown or reboot

**Comment by 1000001101000** [comment]
they are reset to the chip defaults when the chip is powered off or receives a reset command. As long as nothing else comes along and changes them they tend to stay while the system is up but won't persist past a reboot or anything like that.

**Comment by jcmoore1977** [comment]
Thanks, that’s what I am needing to know. Going to try to fix the speed 1000 thing.

**Comment by 1000001101000** [comment]
Have you had any luck?

I tried doing some additional searches working backwards from what "fixed" the problem, trying to find what typical causes might be to and find how those have been solved. Pretty much everything I came across  involved bad Ethernet cables or switches.

I believe you did a good job of accounting for that sort of thing when we first talked about this issue but I'd thought I'd mention it in case you want to try a different Ethernet cable or something.

**Comment by jcmoore1977** [comment]
I am using a brand new cable that came with my other NAS system. I could possibly try hooking it into a different switch but the only other switch I have is a 10/100 base. The one I am using is a 10/100/1000 base.

**Comment by 1000001101000** [comment]
hmm, I wonder if it would auto-negotiate with the 10/100 switch, choose 100M and be stable. It wouldn't tell us much though.

**Comment by 1000001101000** [comment]
I just got done reworking the installer images. I learned a lot from the hard ~7MB size limit that the really old devices had and applied some of that to the more modern devices. As part of that process I also added ethtool to the installer image and configured it to install ethtool in the final image. I haven't done anything with it yet but the groundwork is all in place now

we could add some logic to the script I'm using to set the MAC address when the installer starts to also set the maximum speed to 100M and see if that allows a stable install for these devices. I really doubt that will hurt anyone's experience during the installer. 

As for the installed system we could try creating a daemon that watches for the link up/down messages and sets it to 100M once it encounters a certain number. I'd be somewhat concerned about adding that to all installs since it might have unforeseen false positives we haven't anticipated. We could make enabling it optional, but without a reliable network connection it might be hard to actually enable it.

Out of curiosity, what model switch are you using currently?

**Comment by livstone** [comment]
Hi 1000001101000,
you may know me from [this thread in debianforum.de](https://debianforum.de/forum/viewtopic.php?f=32&t=177887) about emulating armada 370 machines and another thread about building kernel modules. Thanks again for your tips.
Your buster install image worked ootb and fortunately there emerged no network or other problems on my LS441DE box except for a minor issue concerning wrong alarm time settings on rtc_rs5c372.
If I can help on testing some features, please let me know. It would be a pleasure to solve same open issues.

**Comment by 1000001101000** [comment]
Good to hear from you!

One of these days I need to see if I can get debian to add the rtc_rs5c372 to their default and lpae armhf kernels. actually, if you're up for it would be really helpful to have someone start that process for me. 

If you could somehow convince Buffalo to post the u-boot source code for the LS441DE so that I can fix the network issue for this thread that would also be amazing (though possibly not possible). I actually know how to fix the issue now I just haven't been able to build a u-boot that actually works with the LS441DE.

I'll also keep you in mind when I get around to testing Bullseye as it gets closer to release. I can always use additional testers for those types of things.
